# Week7 - Decision Tree Lab

* Train-test split
* Train a decison tree model
* Train a random forest model
* Evaluate the models
* Explain findings

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import  accuracy_score, precision_score, recall_score, f1_score
from sklearn.ensemble import RandomForestClassifier

In [ ]:
dataframe = pd.read_csv('https://raw.githubusercontent.com/msaricaumbc/DS_data/master/ds602/log_reg/employee-turnover-balanced.csv')

In [ ]:
dataframe

left_company  age frequency_of_travel              department  \
0             No   37       Travel_Rarely                   Sales   
1             No   39       Travel_Rarely  Research & Development   
2             No   52   Travel_Frequently  Research & Development   
3             No   50          Non-Travel                   Sales   
4             No   44       Travel_Rarely  Research & Development   
..           ...  ...                 ...                     ...   
995          Yes   39       Travel_Rarely                   Sales   
996          Yes   26       Travel_Rarely                   Sales   
997          Yes   18   Travel_Frequently                   Sales   
998          Yes   28       Travel_Rarely  Research & Development   
999          Yes   29   Travel_Frequently                   Sales   

     commuting_distance  education  satisfaction_with_environment  gender  \
0                    16          4                              4    Male   
1                     3          2                              3    Male   
2                    25          4                              3  Female   
3                     1          3                              4  Female   
4                     4          3                              4    Male   
..                  ...        ...                            ...     ...   
995                   3          2                              4  Female   
996                   4          4                              4    Male   
997                   5          3                              2    Male   
998                   2          4                              1    Male   
999                  13          3                              1  Female   

     seniority_level                   position  satisfaction_with_job  \
0                  2            Sales Executive                      3   
1                  2      Laboratory Technician                      3   
2                  4     Manufacturing Director                      4   
3                  2            Sales Executive                      3   
4                  2  Healthcare Representative                      2   
..               ...                        ...                    ...   
995                2            Sales Executive                      3   
996                2            Sales Executive                      4   
997                1       Sales Representative                      2   
998                1         Research Scientist                      4   
999                2            Sales Executive                      2   

    married_or_single  last_raise_pct  last_performance_rating  \
0            Divorced              19                        3   
1            Divorced              15                        3   
2             Married              22                        4   
3             Married              12                        3   
4              Single              12                        3   
..                ...             ...                      ...   
995           Married              18                        3   
996            Single              12                        3   
997            Single              14                        3   
998           Married              13                        3   
999            Single              11                        3   

     total_years_working  years_at_company  years_in_current_job  \
0                      9                 1                     0   
1                     11                10                     8   
2                     31                 9                     8   
3                     19                18                     7   
4                     10                 5                     2   
..                   ...               ...                   ...   
995                   12                 1                     0   
996                    8

In [ ]:
features = dataframe[['department', 'position', 'education', 'gender', 'married_or_single', 'seniority_level', 'age', 'total_years_working', 'years_at_company', 'years_in_current_job', 'years_since_last_promotion', 'years_with_current_supervisor', 'last_raise_pct', 'last_performance_rating', 'frequency_of_travel', 'commuting_distance', 'satisfaction_with_job', 'satisfaction_with_environment']]
label = dataframe['left_company']

In [ ]:

features = pd.get_dummies(features, drop_first=True)

In [ ]:

X_train, X_test, y_train, y_test = train_test_split(features, label, test_size=0.2, random_state=1)

In [ ]:
#Train a decision tree model
decision_tree = DecisionTreeClassifier()
decision_tree.fit(X_train, y_train)

DecisionTreeClassifier()

In [ ]:
random_forest = RandomForestClassifier(n_estimators=100)
random_forest.fit(X_train, y_train)

# Use the trained model to make predictions on the testing data

y_pred_dt = decision_tree.predict(X_test)
y_pred_rf = random_forest.predict(X_test)

#Evaluation Metrics of Decision Tree

acc_decision_tree = accuracy_score(y_test, y_pred_dt)
print("Accuracy of Decision Tree:", acc_decision_tree)


acc_rf = accuracy_score(y_test, y_pred_rf)
print("Random Forest model accuracy:", acc_rf)

Accuracy of Decision Tree: 0.77
Random Forest model accuracy: 0.865


The Random Forest model achieved a higher accuracy score of 0.865 compared to the decision tree model's accuracy score of 0.79. This suggests that the Random Forest model outperformed the decision tree model in terms of making accurate predictions on the test dataset.